EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs

```diff
@@ -54,39 +54,47 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
                     continue;
                 }
 
-                var enumValues = Enum.GetValuesAsUnderlyingType(propertyType).Cast<object>().Distinct().ToArray();
-                if (enumValues.Length == 0)
+                if (!_cachedConstraints.TryGetValue(propertyType, out var constraintSql))
                 {
-                    continue;
-                }
-
-                sql.Clear();
+                    var enumValues = Enum.GetValuesAsUnderlyingType(propertyType).Cast<object>().Distinct().ToArray();
+                    if (enumValues.Length == 0)
+                    {
+                        continue;
+                    }
 
-                sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+                    sql.Clear();
 
-                if (enumValues.Length > 2
-                    && TryParseContiguousRange(typeMapping.Converter, enumValues, out var minValue, out var maxValue))
-                {
-                    sql.Append(" BETWEEN ");
-                    sql.Append(typeMapping.GenerateSqlLiteral(minValue));
-                    sql.Append(" AND ");
-                    sql.Append(typeMapping.GenerateSqlLiteral(maxValue));
-                }
-                else
-                {
-                    sql.Append(" IN (");
-                    foreach (var item in enumValues)
+                    if (enumValues.Length > 2
+                        && TryParseContiguousRange(typeMapping.Converter, enumValues, out var minValue, out var maxValue))
+                    {
+                        sql.Append(" BETWEEN ");
+                        sql.Append(typeMapping.GenerateSqlLiteral(minValue));
+                        sql.Append(" AND ");
+                        sql.Append(typeMapping.GenerateSqlLiteral(maxValue));
+                    }
+                    else
                     {
-                        var value = typeMapping.GenerateSqlLiteral(item);
-                        sql.Append($"{value}, ");
+                        sql.Append(" IN (");
+                        foreach (var item in enumValues)
+                        {
+                            var value = typeMapping.GenerateSqlLiteral(item);
+                            sql.Append($"{value}, ");
+                        }
+
+                        sql.Remove(sql.Length - 2, 2);
+                        sql.Append(')');
                     }
 
-                    sql.Remove(sql.Length - 2, 2);
-                    sql.Append(')');
+                    constraintSql = sql.ToString();
                 }
 
-                var constraintName = $"CK_{tableName}_{columnName}_Enum";
-                entityType.AddCheckConstraint(constraintName, sql.ToString());
+                entityType.AddCheckConstraint(
+                    $"CK_{tableName}_{columnName}_Enum",
+                    sql
```

Comment:
This would probably only be a difference of a few milliseconds, but I feel like we should be able to _not_ have to clear and re-add to the sql string builder.

Maybe something like this pseudocode:

```
length is 0? continue
sql.Clear().Append(columnName)
if cached version exists
  sql.Append(cachedCheck)
otherwise
  generate, cache, and append generated check
entityType.AddCheckConstraint(name, sql.ToString())
```

Does that make sense?

Reply:
Yeah, I see what you mean... Though that extra Clear only happens when we see a new enum type; for cases where we already have the cached constraint, we just Clear once. Given that StringBuilder.Clear is an extremely light operation, I'm not sure it's worth it.

Reply:
Now that you've moved the generation logic into its own method, this is a moot point. Agree it's very light, but the fastest code is the code we don't have to run!